### ISSUE TITLE
libudev running  inside a netns is not able to handle network device correctly

### ISSUE THREAD

**Issue by pespin** [systemd/systemd#6713]
### Submission type

 - Bug report

### systemd version the issue has been seen with

> core/libsystemd 234.11-8 from Archlinux repos

### Used distribution

> Archlinux

### In case of bug report: Expected behaviour you didn't see

> libudev failing to find iface in netns while used in a process running inside the netns.

### In case of bug report: Unexpected behaviour you saw

> See below.

### In case of bug report: Steps to reproduce the problem

> 
I have a GSM modem connected to my PC running systemd-udev, which exposes a network interface wwp0s29u1u2 which can be used to send data over GPRS. I use ofono to manage the modem, which uses libgudev (-> libudev) to find the modems in the system.

For routing purposes, I want/need to run ofono + modem interface (wwp0s29u1u2) on a network namespace different than the default one.

I create a network namespace: sudo ip netns add myns
I move the iface to the network namespace: sudo ip link set dev wwp0s29u1u2 netns myns
I start ofono in the network namespace: sudo ip netns exec myns strace ofonod -n -d

When ofono at startup uses libudev, it doesn't find the wwp0s29u1u2 interface from the USB modem device, despite it being reachable because it's on the same network namespace. All the /sys/* paths seem to be working, fine, but the issue comes when it tries to read "/run/udev/data/n6" (from IFINDEX=6 in /sys/devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.2/4-1.2:1.0/net/wwp0s29u1u2/uevent I guess). However, that file (contents attached here) is not there anymore since the time I moved the interface wwp0s29u1u2 to the netns. When I move the interface back to the default netns, the file /run/udev/data/n6 appears again.

output from strace while running ofono using libudev:
```
15:11:53 access("/sys/devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.2/4-1.2:1.0/net/wwp0s29u1u2/uevent", F_OK) = 0
15:11:53 open("/sys/devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.2/4-1.2:1.0/net/wwp0s29u1u2/uevent", O_RDONLY|O_CLOEXEC) = 11
15:11:53 fstat(11, {st_mode=S_IFREG|0644, st_size=4096, ...}) = 0
15:11:53 fstat(11, {st_mode=S_IFREG|0644, st_size=4096, ...}) = 0
15:11:53 read(11, "DEVTYPE=wwan\nINTERFACE=wwp0s29u1"..., 4096) = 45
15:11:53 read(11, "", 4096)             = 0
15:11:53 close(11)                      = 0
15:11:53 readlinkat(AT_FDCWD, "/sys/devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.2/4-1.2:1.0/net/wwp0s29u1u2/subsystem", "../../../../../../../../../class"..., 99) = 36
15:11:53 getrandom("\x3a\x11\x99\x92\xbe\xd4\x22\x47\x85\xdd\x73\xa3\xfb\x50\x12\xe1", 16, GRND_NONBLOCK) = 16
15:11:53 open("/run/udev/data/n6", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
15:11:53 getdents(10, /* 0 entries */, 32768) = 0
15:11:53 close(10)                      = 0
```

Content of /run/udev/data/n6 (when present). I suspect ofono needs any of the information there or it fails because libudev returns an error?
```
I:22335603687
E:DEVPATH=/devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.2/4-1.2:1.0/net/wwp0s29u1u2
E:SUBSYSTEM=net
E:DEVTYPE=wwan
E:INTERFACE=wwp0s29u1u2
E:IFINDEX=6
E:USEC_INITIALIZED=22335603687
E:ID_NET_NAME_PATH=wwp0s29u1u2
E:ID_VENDOR=Qualcomm_Incorporated
E:ID_VENDOR_ENC=Qualcomm\x20Incorporated
E:ID_VENDOR_ID=05c6
E:ID_MODEL=Qualcomm_Gobi_2000
E:ID_MODEL_ENC=Qualcomm\x20Gobi\x202000
E:ID_MODEL_ID=9205
E:ID_REVISION=0002
E:ID_SERIAL=Qualcomm_Incorporated_Qualcomm_Gobi_2000
E:ID_TYPE=generic
E:ID_BUS=usb
E:ID_USB_INTERFACES=:ffffff:
E:ID_USB_INTERFACE_NUM=00
E:ID_USB_DRIVER=qmi_wwan
E:ID_VENDOR_FROM_DATABASE=Qualcomm, Inc.
E:ID_MODEL_FROM_DATABASE=Gobi 2000
E:ID_MM_CANDIDATE=1
E:ID_PATH=pci-0000:00:1d.0-usb-0:1.2:1.0
E:ID_PATH_TAG=pci-0000_00_1d_0-usb-0_1_2_1_0
E:ID_NET_DRIVER=qmi_wwan
E:ID_NET_LINK_FILE=/usr/lib/systemd/network/99-default.link
E:SYSTEMD_ALIAS=/sys/subsystem/net/devices/wwp0s29u1u2 /sys/subsystem/net/devices/wwp0s29u1u2
E:TAGS=:systemd:
G:systemd
```

While running "udevadm monitor --subsystem=net -k" both inside the netns and outside the netns, I can see the interface moving at the time I use ip link to move the iface inside the netns:

Outside the netns (default netns):
```
$ udevadm monitor --subsystem=net -k
monitor will print the received events for:
KERNEL - the kernel uevent

KERNEL[18385.315169] remove   /devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.2/4-1.2:1.0/net/wwp0s29u1u2 (net)
```

Inside the netns:
```
# udevadm monitor --subsystem=net -k
monitor will print the received events for:
KERNEL - the kernel uevent

KERNEL[18385.315376] add      /devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.2/4-1.2:1.0/net/wwp0s29u1u2 (net)
KERNEL[18385.315620] move     /devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.2/4-1.2:1.0/net/wwp0s29u1u2 (net)
```

So, basically everything seems fine until libudev tries to use its own files (/run/udev/data/n6). As far as I understand, that file should be at least available to processes running inside netns, perhaps by means like bind explained in "ip netns exec" http://man7.org/linux/man-pages/man8/ip-netns.8.html. Other possibility would be that libudev tries harder to generate the content of the /run/udev/data/n6 if it is not found, because afaik from that process point of view all information should be there?

**Comment by poettering** [comment]
Sorry, there's nothing we can do about that: AF_NETLINK is covered by netns, too, and AF_NETLINK is what libudev uses internally. Hence, if you open your own network namespace you lose udev support. And that's not a limitation of libudev really, it's the limiation of the Linux kernel. Sorry!

**Comment by laf0rge** [comment]
Hi Lennart, thanks for the quick follow-up.
I'm still wondering what exactly libudev is using "internally" that will no longer work when a new AF_NETLINK "instance" is created in the new network namespace.  It would be great if you could elaborate  a bit more on that.  According to the bug report, the library simply wants to open a text file, and nothing that relates to AF_NETLINK at that time?

**Comment by poettering** [comment]
udev and libudev get their notifications about devices coming/going and changing from the kernel via an AF_NETLINK socket. It's a special communication channel between the kernel and udev.
